server/src/test/java/e2e/ChildThreadTest.java

```diff
@@ -0,0 +1,199 @@
+ *
+ * For backwards compatibility, we will always have the option to have un-typed
+ * ExternalEventDef's, so I think that is safe. However, we will encourage our users
+ * to always use strong typing.
+ */
+@LHTest(externalEventNames = {ChildThreadTest.COMPLETE_CHILD, ChildThreadTest.COMPLETE_PARENT})
+public class ChildThreadTest {
+
+    public static final String COMPLETE_CHILD = "agjo-complete-child";
+    public static final String COMPLETE_PARENT = "agjo-complete-parent";
+
+    private WorkflowVerifier verifier;
+
+    @LHWorkflow("basic-child-thread")
+    private Workflow basicChildThread;
+
+    @Test
+    void childCompletesFirstAndCanMutateParentVariables() {
+        verifier.prepareRun(basicChildThread)
+                .thenVerifyWfRun(wfRun -> {
+                    // The child should be created in the first Command
+                    Assertions.assertThat(wfRun.getThreadRuns(1).getStatus()).isEqualTo(LHStatus.RUNNING);
+                })
+                .thenSendExternalEventWithContent(COMPLETE_CHILD, Map.of("result", 1234))
+                .thenVerifyVariable(0, "shared-var", value -> {
+                    Assertions.assertThat(value.getInt()).isEqualTo(1234);
+                })
+                .thenVerifyWfRun(wfRun -> {
+                    Assertions.assertThat(wfRun.getThreadRuns(1).getStatus()).isEqualTo(LHStatus.COMPLETED);
+                })
+                .thenSendExternalEventWithContent(COMPLETE_PARENT, Map.of("result", 4321))
+                .thenVerifyNodeRun(0, 3, nodeRun -> {
+                    Assertions.assertThat(nodeRun.getNodeTypeCase()).isEqualTo(NodeTypeCase.WAIT_THREADS);
+                    WaitForThreadsRun wftr = nodeRun.getWaitThreads();
+
+                    Assertions.assertThat(wftr.getThreadsCount()).isEqualTo(1);
+                    WaitForThread childStatus = wftr.getThreads(0);
+                    Assertions.assertThat(childStatus.getThreadRunNumber()).isEqualTo(1);
+                    Assertions.assertThat(childStatus.getThreadStatus()).isEqualTo(LHStatus.COMPLETED);
+                })
+                .start();
+    }
+
+    @Test
+    void childErrorPropagatesToParentOnlyOnWaitForThreadsNode() {
+        verifier.prepareRun(basicChildThread)
+                .thenVerifyWfRun(wfRun -> {
+                    // The child should be created in the first Command
+                    Assertions.assertThat(wfRun.getThreadRuns(1).getStatus()).isEqualTo(LHStatus.RUNNING);
+                })
+                .thenSendExternalEventWithContent(COMPLETE_CHILD, Map.of("result", "not-an-integer"))
+                .thenVerifyNodeRun(1, 1, nodeRun -> {
+                    // I thought this should be true, but apparently it's not. We need to review this.
+                    // Assertions.assertThat(nodeRun.getStatus()).isEqualTo(LHStatus.ERROR);
+                    Assertions.assertThat(nodeRun.getFailuresCount()).isEqualTo(1);
```

Comment:
I think I found a problem, or at least behavior that is different than what I expected. I had a NodeRun whose SubNodeRun completed properly but the variable mutations failed. The result is:
- :smiley: The ThreadRun is in the ERROR state
- :thinking: There is a Failure with type VAR_MUTATION_ERROR on the NodeRun rather than VAR_MUTATION_ERROR
- :thinking_face: The NodeRun is in the status COMPLETED

I thought that we decided that since an Edge is part of a Node, and the VariableMutation lives on the Edge, then technically the VariableMutation is part of the NodeRun itself. So therefore the NodeRun should be in the ERROR status...what do you think?
My tests show that this is not necessarily the case. Maybe I need to review that PR I made in March which fixes this 

(Link to original slack message, private to LH employees: https://littlehorseworkspace.slack.com/archives/C060A2XCX9D/p1725819836840409)


[11:27](https://littlehorseworkspace.slack.com/archives/C060A2XCX9D/p1725820034527379)
Also, another problem is that the failure is of the type VAR_SUB_ERROR not VAR_MUTATION_ERROR. I don't know if we actually use the status VAR_MUTATION_ERROR. Maybe we can get rid of it?

Reply:
Hazim had a similar conversation a few months ago, in which he mentioned that the LH server handles mutation failures in a confusing way. Given the current design, I believe this is the best mechanism for handling variable mutation failures, as the node itself didn't fail. For instance, consider a wfspec that defines a TaskNode, defines several other variables, and then mutates those variables.

```
wf.execute('my-silly-validations');

wf.addVariable(...)
// More variables here

wf.mutate(...)
// More mutations here
```

Debugging will be quite confusing at first, clients may believe that the 'my-silly-validations' node has failed, which isn't the case. Therefore, in this scenario, it's preferable to have an edge failure within the node, indicating that the node should be completed. If we really want to change this behavior, I recommend splitting nodes and edges into separate thread-level objects.

Reply:
There is a difference between the `SubNodeRun` and the `NodeRun`. I think the fact that the `NodeRun` has a `Failure` on it and yet its status is `COMPLETED` is bad. Especially because the `Mutation` is part of the `Node`